### PR TITLE
hack: helper scripts and testing via Dockerfile

### PR DESCRIPTION
**Pull request by tonistiigi**

### CODE REVIEW

**Comment by tonistiigi** [comment]
GH actions coming as follow-up https://github.com/tonistiigi/fsutil/actions/runs/174361244

File: receive_test.go [hunk 1]
```diff
@@ -337,18 +337,19 @@ file zzz.aa
...
-func sockPair(ctx context.Context) (Stream, Stream) {
-	c1 := make(chan *types.Packet, 32)
-	c2 := make(chan *types.Packet, 32)
-	return &fakeConn{ctx, c1, c2}, &fakeConn{ctx, c2, c1}
-}
+// func sockPair(ctx context.Context) (Stream, Stream) {
+// 	c1 := make(chan *types.Packet, 32)
+// 	c2 := make(chan *types.Packet, 32)
+// 	return &fakeConn{ctx, c1, c2}, &fakeConn{ctx, c2, c1}
+// }
```
Comment (AkihiroSuda) [inline receive_test.go:344]: Can we remove unused funcs?

Comment (tonistiigi) [reply receive_test.go]: removed this. left in some comments in benchmark for debug reference

File: bench/bench_test.go [hunk 2]
```diff
@@ -162,6 +157,9 @@ func BenchmarkCopyWithTar1000(b *testing.B) {
 	benchmarkInitialCopy(b, copyWithTar, 1000)
 }
 
+// func chrootCopyWithTar(src, dest string) error {
```
Comment (thaJeztah) [inline bench/bench_test.go:160]: You left these (and below) on purpose?

Comment (tonistiigi) [reply bench/bench_test.go:160]: yes

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    go test -test.v ./...
+
+FROM base AS test-noroot
+RUN mkdir /go/pkg && chmod 0777 /go/pkg
+USER 1000:1000
+RUN --mount=target=. \
+    --mount=target=/tmp/.cache,type=cache \
+    GOCACHE=/tmp/gocache go test -test.v ./...
+
+FROM build
```
Comment (thaJeztah) [inline Dockerfile:29]: Missing newline here as well

File: hack/lint [hunk 4]
```diff
@@ -8,5 +8,8 @@ set -eu -o pipefail -x
 progressFlag=""
 if [ "$CONTINUOUS_INTEGRATION" == "true" ]; then progressFlag="--progress=plain"; fi
 
+cacheOnlyFlag=""
+if ! docker build 2>&1 --help | grep buildx > /dev/null; then cacheOnlyFlag="-o type=cacheonly"; fi
+
 export DOCKER_BUILDKIT=1
-docker build -o type=cacheonly -f ./hack/dockerfiles/lint.Dockerfile .
\ No newline at end of file
+docker build $progressFlag $cacheOnlyFlag -f ./hack/dockerfiles/lint.Dockerfile .
```
Comment (thaJeztah) [inline hack/lint:15]: And here 😅

File: .travis.yml [hunk 5]
```diff
@@ -1,24 +1,26 @@
...
- - sudo ./test -test.v
- - GOOS=darwin go build ./...
- - GOOS=windows go build ./...
- - GOARCH=arm GOARM=7 go build ./...
+  - docker build --progress=plain --build-arg=GO_VERSION .
+  - docker build --progress=plain --build-arg=GO_VERSION --target=test .
+  - docker build --progress=plain --build-arg=GO_VERSION --target=test-noroot .
+  - ./hack/validate-gomod
+  - ./hack/lint
+  - docker build --progress=plain --build-arg=GO_VERSION --platform=linux/arm64,darwin/amd64,windows/amd64,linux/arm,linux/arm64 .
```
[the remark this answers is not in the document]
Comment (tonistiigi) [reply .travis.yml:26]: saved a whole byte

File: hack/validate-gomod [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+set -eu
+
+: ${CONTINUOUS_INTEGRATION=}
+: ${DOCKER_BUILDKIT=}
+
+progressFlag=""
+if [ "$CONTINUOUS_INTEGRATION" = "true" ]; then progressFlag="--progress=plain"; fi
+
+cacheOnlyFlag=""
+if ! docker build 2>&1 --help | grep buildx > /dev/null; then cacheOnlyFlag="-o type=cacheonly"; fi
```
Comment (thaJeztah) [inline hack/validate-gomod:12]: (`shfmt` changes this to)

```suggestion
if ! docker build --help 2>&1 | grep buildx >/dev/null; then cacheOnlyFlag="-o type=cacheonly"; fi
```

File: hack/validate-gomod [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  export DOCKER_BUILDKIT=1
+  docker build $progressFlag $cacheOnlyFlag -f ./hack/dockerfiles/gomod.Dockerfile --target validate . || exit 1
+  ;;
+check)
+  status="$(git status --porcelain -- go.mod go.sum 2>/dev/null)"
+  diffs=$(echo "$status" | grep -v '^[RAD] ' || true)
+  if [ "$diffs" ]; then
+      {
+	set +x
+        echo 'The result of "./hack/validate-gomod" differs'
```
Comment (thaJeztah) [inline hack/validate-gomod:25]: Looks like you're using spaces for indentation; could you run `shfmt -w .` (using https://github.com/mvdan/sh)?

File: .travis.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (thaJeztah) [reply .travis.yml:26]: 😂 yeah, well, there _are_ some valid reasons https://unix.stackexchange.com/a/18789

**Comment by tonistiigi** [comment]
@thaJeztah added shfmt to ci

**Pull request merged**